knora/excel2xml.py

```diff
@@ -0,0 +1,2127 @@
+        value_ = etree.Element(
+            "{%s}time" % (xml_namespace_map[None]),
+            **kwargs,
+            nsmap=xml_namespace_map
+        )
+        value_.text = val.value
+        prop_.append(value_)
+
+    return prop_
+
+
+def make_uri_prop(
+    name: str,
+    value: Optional[Union[PropertyElement, str, Any, Iterable[Union[PropertyElement, str, Any]]]] = None,
+    values: Optional[Iterable[Union[PropertyElement, str, Any]]] = None,
+    calling_resource: str = ""
+) -> etree._Element:
+    """
+    Make a <uri-prop> from one or more URIs. The URI(s) can be provided as string or as PropertyElement with a string
+    inside. If provided as string, the permission for every value is "prop-default".
+
+    To create one ``<uri>`` child, use the param ``value``, to create more than one ``<uri>`` children, use ``values``.
+
+    Args:
+        name: the name of this property as defined in the onto
+        value: a string/PropertyElement, or an iterable of identical strings/PropertyElements
+        values: an iterable of (usually distinct) strings/PropertyElements
+        calling_resource: the name of the parent resource (for better error messages)
+
+    Examples:
+        >>> make_uri_prop(":testproperty", "www.test.com")
+                <uri-prop name=":testproperty">
+                    <uri permissions="prop-default">www.test.com</uri>
+                </uri-prop>
+        >>> make_uri_prop(":testproperty", PropertyElement("www.test.com", permissions="prop-restricted", comment="example"))
+                <uri-prop name=":testproperty">
+                    <uri permissions="prop-restricted" comment="example">www.test.com</uri>
+                </uri-prop>
+        >>> make_uri_prop(":testproperty", values=["www.1.com", "www.2.com"])
+                <uri-prop name=":testproperty">
+                    <uri permissions="prop-default">www.1.com</uri>
+                    <uri permissions="prop-default">www.2.com</uri>
+                </uri-prop>
+        >>> make_uri_prop(":testproperty", value=["www.1.com", "www.2.com"])
+                ERROR. Use this if some fields in your data source should be equal, but you cannot trust
+                your source. This method call will only work if all items of "value" are identical.
+
+    See https://docs.dasch.swiss/latest/DSP-TOOLS/dsp-tools-xmlupload/#uri-prop
+    """
+
+    # check the input: prepare a list with valid values
+    values_new = _check_and_prepare_values(
+        value=value,
+        values=values,
+        name=name,
+        calling_resource=calling_resource
+    )
+
+    # check value type
+    for val in values_new:
+        # URI = scheme ":" ["//" host [":" port]] path ["?" query] ["#" fragment]
+        if not regex.search(
+            r"(?<scheme>[a-z][a-z0-9+.\-]*):(//(?<host>[\w_.\-\[\]:~]+)(?<port>:\d{0,6})?)(?<path>/[\p{L}_\-.~]*)*"
+            r"(?<query>\?[\p{L}_.\-=]+)*(?<fragment>#[\p{L}_/\-~:.]*)?", str(val.value), flags=regex.UNICODE):
```

Comment:
seeing how many Regexes we have floating around, maybe they should all be collected in one file (and get pre-compiled). to ensure that touching a regex doesn't mean having to find all the places in the code where this regex appears, and changing it everywhere.

Reply:
Noted for a next task.